modules/internal/DefaultAssociative.chpl

```diff
@@ -566,14 +566,15 @@ module DefaultAssociative {
 
     // ref version
     proc dsiAccess(idx : idxType, haveLock = false) ref {
-      const shouldLock = dom.parSafe && !haveLock;
+      const arrOwnsDom = dom._arrs.length == 1;
+      const shouldLock = dom.parSafe && !haveLock && arrOwnsDom;
```

Comment:
To me, your new example seems to be across the line where the user is doing enough disparate operations on a single associative array that they should be expected to handle any locking required to make it safe themselves.  Even the 'if A.domain.member(key)' check seems suspect since there's a potential race (what if it is a member by the time they're done checking it?).